eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java

```diff
@@ -0,0 +1,121 @@
+    private boolean hasServiceClientBuilderAnnotation;
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.CLASS_DEF,
+            TokenTypes.METHOD_DEF
+        };
+    }
+
+    @Override
+    public void leaveToken(DetailAST token) {
+        switch (token.getType()) {
+            case TokenTypes.CLASS_DEF:
+                // end of CLASS_DEF node, reset the value back to previous state
+                hasServiceClientBuilderAnnotation = hasServiceClientBuilderAnnotationStack.pop();
+                break;
+            default:
+                // Checkstyle complains if there's no default block in switch
+                break;
+        }
+    }
+
+    @Override
+    public void visitToken(DetailAST token) {
+        switch (token.getType()) {
+            case TokenTypes.CLASS_DEF:
+                // Save the state of variable 'hasServiceClientBuilderAnnotation' to limit the scope of accessibility
+                hasServiceClientBuilderAnnotationStack.push(hasServiceClientBuilderAnnotation);
+
+                final DetailAST serviceClientAnnotationBuilderToken = getServiceClientBuilderAnnotation(token);
+                final String className = token.findFirstToken(TokenTypes.IDENT).getText();
+
+                hasServiceClientBuilderAnnotation = serviceClientAnnotationBuilderToken != null;
+                if (hasServiceClientBuilderAnnotation) {
+                    // Don't need to check if the 'serviceClients' exist. It is required when using @ServiceClientBuilder
+
+                    // HAS @ServiceClientBuilder annotation but NOT named the class <ServiceName>ClientBuilder
+                    if (!className.endsWith("ClientBuilder")) {
+                        log(token, String.format("Class annotated with @ServiceClientBuilder ''%s'' should be named <ServiceName>ClientBuilder.", className));
+                    }
+                } else {
+                    // No @ServiceClientBuilder annotation but HAS named the class <ServiceName>ClientBuilder
+                    if (className.endsWith("ClientBuilder")) {
+                        log(token, String.format("Class ''%s'' should be annotated with @ServiceClientBuilder.", className));
+                    }
+                }
+                break;
+            case TokenTypes.METHOD_DEF:
+                if (!hasServiceClientBuilderAnnotation) {
+                    return;
+                }
+
+                final String methodName = token.findFirstToken(TokenTypes.IDENT).getText();
+                // method name has prefix 'build' but not 'buildClient' or 'buildAsyncClient'
+                if (methodName.startsWith("build") && !BUILD_ASYNC_CLIENT.equals(methodName) && !BUILD_CLIENT.equals(methodName)) {
```

Comment:
There should also be a check to ensure there is at least one method that starts with "build".